### PR TITLE
Detect possible async calls in scripts

### PR DESCRIPTION
**Pull request by dranikpg**
Regex for detecting possible async calls for scripts

### CODE REVIEW

**Comment by dranikpg** [comment]
@romange PTAL and maybe you spot some bugs or missing cases

I just found one: a multi-line expression can be divided by multiple comment-only lines that we need to exclude.

**Comment by dranikpg** [comment]
I thought maybe we can just detect all comments ahead and remove them - it makes things much easier, however it wouldn't preserve the line numbers for error reporting. Instead of removing then we could replace then with some placeholder, but it doesn't make much sense as the regex still needs to account for them, just in a simper way

File: src/server/script_mgr.cc [hunk 1]
```diff
@@ -202,6 +202,10 @@ io::Result<string, GenericError> ScriptMgr::Insert(string_view body, Interpreter
   if (!params)
     return params.get_unexpected();
 
+  auto async_body = Interpreter::DetectPossibleAsyncCalls(body);
```
Comment (romange) [inline src/server/script_mgr.cc:205]: I think it's better we keep both the original and the processed versions. 
we have "script list" that can help us to display both and then we will be able to debug possible bugs in prod

**Comment by romange** [comment]
The question is whether the frameworks we care about use these hard cases or not. 
I think we can be conservative about the cases we want to support, i.e. if it's easier to recognize and skip hard cases then  lets do that instead of finding a way to replace them.

**Comment by romange** [comment]
Another thing I just thought is that it's hard to explain "acall" semantics to users because "asynchronicity"  is the implementation detail in Dragonfly. What we really want to say, IMHO, is that a call that explicitly does not return any results is more efficient in Dragonfly.

In that case, should we call it "proc" and not "acall"  ?

**Comment by dranikpg** [comment]
> In that case, should we call it "proc" and not "acall" ?

Yes, we can call it proc and ppoc(?). I can do a renaming PR afterwards, but lets keep it out of scope here

**Comment by romange** [comment]
What's a ppoc?

**Comment by dranikpg** [comment]
*pproc - pcall but without reply

**Comment by dranikpg** [comment]
Maybe we should enable it only with a flag? See #1091

**Comment by romange** [comment]
The main use-case is for atomic scripts.
Yes, setting the flag is a good idea, but I think it should be enabled by
default.

File: src/server/script_mgr.cc [hunk 2]
```diff
@@ -192,28 +194,47 @@ io::Result<optional<ScriptMgr::ScriptParams>, GenericError> DeduceParams(string_
...
+  if (!params_opt)
+    return params_opt.get_unexpected();
+  auto params = params_opt->value_or(default_params_);
+
+  // If the script is atomic, check for possible squashing optimizations.
+  // For non atomic modes, squashing increases the time locks are held, which
+  // can decrease throughput with frequently accessed keys.
+  optional<string> async_body;
+  if (params.atomic) {
+    if (async_body = Interpreter::DetectPossibleAsyncCalls(body); async_body)
```
Comment (romange) [inline src/server/script_mgr.cc:222]: following what we discussed, please put it under flag. I am fine with flipping the flag to on but lets have some guard.

**Review by romange — APPROVED** [review]

**Pull request merged**